File: src/CSharp/CodeCracker/Usage/UriAnalyzer.cs
using CodeCracker.CSharp.Usage.MethodAnalyzers;
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.Diagnostics;
using System;
using System.Collections.Immutable;

namespace CodeCracker.CSharp.Usage
{
    [DiagnosticAnalyzer(LanguageNames.CSharp)]
    public class UriAnalyzer : DiagnosticAnalyzer
    {
        internal const string Title = "Your Uri syntax is wrong.";
        internal const string MessageFormat = "{0}";
        internal const string Category = SupportedCategories.Usage;

        private const string Description = "This diagnostic checks the Uri string and triggers if the parsing fail "
                                           + "by throwing an exception.";

        internal static readonly DiagnosticDescriptor Rule = new DiagnosticDescriptor(
            DiagnosticId.Uri.ToDiagnosticId(),
            Title,
            MessageFormat,
            Category,
            DiagnosticSeverity.Error,
            isEnabledByDefault: true,
            description: Description,
            helpLinkUri: HelpLink.ForDiagnostic(DiagnosticId.Uri));

        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);

        public override void Initialize(AnalysisContext context) => context.RegisterSyntaxNodeAction(Analyzer, SyntaxKind.ObjectCreationExpression);

        private static void Analyzer(SyntaxNodeAnalysisContext context)
        {
            if (context.IsGenerated()) return;
            var mainConstrutor = new MethodInformation(
                nameof(Uri),
                "System.Uri.Uri(string)",
                args =>
                {
                    {
                        if (args[0] == null)
                        {
                            return;
                        }
                        new Uri(args[0].ToString());
                    }
                }
            );
            var constructorWithUriKind = new MethodInformation(
                nameof(Uri),
                "System.Uri.Uri(string, System.UriKind)",
                args =>
                {
                    if (args[0] == null)
                    {
                        return;
                    }
                    new Uri(args[0].ToString(), (UriKind)args[1]);
                }
            );

            var checker = new MethodChecker(context, Rule);
            checker.AnalyzeConstructor(mainConstrutor);
            checker.AnalyzeConstructor(constructorWithUriKind);
        }
    }
}

File: src/VisualBasic/CodeCracker/Usage/UriAnalyzer.vb
Imports CodeCracker.VisualBasic.Usage.MethodAnalyzers
Imports Microsoft.CodeAnalysis
Imports Microsoft.CodeAnalysis.VisualBasic
Imports Microsoft.CodeAnalysis.Diagnostics
Imports System.Collections.Immutable

Namespace Usage
    <DiagnosticAnalyzer(LanguageNames.VisualBasic)>
    Public Class UriAnalyzer
        Inherits DiagnosticAnalyzer
        Friend Const Title As String = "Your Uri syntax is wrong."
        Friend Const MessageFormat As String = "{0}"
        Friend Const Category As String = SupportedCategories.Usage

        Private Const Description As String = "This diagnostic checks the Uri string and triggers if the parsing fail " + "by throwing an exception."

        Friend Shared Rule As New DiagnosticDescriptor(DiagnosticId.Uri.ToDiagnosticId(), Title, MessageFormat, Category, DiagnosticSeverity.[Error], isEnabledByDefault:=True,
            description:=Description, helpLinkUri:=HelpLink.ForDiagnostic(DiagnosticId.Uri))

        Public Overrides ReadOnly Property SupportedDiagnostics() As ImmutableArray(Of DiagnosticDescriptor)
            Get
                Return ImmutableArray.Create(Rule)
            End Get
        End Property

        Public Overrides Sub Initialize(context As AnalysisContext)
            context.RegisterSyntaxNodeAction(AddressOf Analyzer, SyntaxKind.ObjectCreationExpression)
        End Sub

        Private Sub Analyzer(context As SyntaxNodeAnalysisContext)
            If context.IsGenerated() Then Return
            Dim mainConstrutor = New MethodInformation("Uri",
                                                       "Public Overloads Sub New(uriString As String)",
                                                       Sub(args)
                                                           If args(0) Is Nothing Then Return
                                                           Dim a = New Uri(args(0).ToString())
                                                       End Sub)
            Dim constructorWithUriKind = New MethodInformation("Uri",
                                                               "Public Overloads Sub New(uriString As String, uriKind As System.UriKind)",
                                                               Sub(args)
                                                                   If args(0) Is Nothing Then Return
                                                                   Dim a = New Uri(args(0).ToString(), DirectCast(args(1), UriKind))
                                                               End Sub)
            Dim checker = New MethodChecker(context, Rule)
            checker.AnalyzeConstructor(mainConstrutor)
            checker.AnalyzeConstructor(constructorWithUriKind)
        End Sub
    End Class
End Namespace